src/Aspire.Hosting/AspireStore.cs

```diff
@@ -0,0 +1,234 @@
+{
+    internal const string AspireStorePathKeyName = "Aspire:Store:Path";
+
+    private readonly string _basePath;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="AspireStore"/> class with the specified base path.
+    /// </summary>
+    /// <param name="basePath">The base path for the store.</param>
+    private AspireStore(string basePath)
+    {
+        ArgumentNullException.ThrowIfNull(basePath);
+
+        _basePath = basePath;
+        EnsureDirectory();
+    }
+
+    /// <summary>
+    /// Gets the base path of this store.
+    /// </summary>
+    public string BasePath => _basePath;
+
+    /// <summary>
+    /// Creates a new instance of <see cref="AspireStore"/> using the provided <paramref name="builder"/>.
+    /// </summary>
+    /// <param name="builder">The <see cref="IDistributedApplicationBuilder"/>.</param>
+    /// <returns>A new instance of <see cref="AspireStore"/>.</returns>
+    public static AspireStore Create(IDistributedApplicationBuilder builder)
+    {
+        ArgumentNullException.ThrowIfNull(builder);
+
+        var assemblyMetadata = builder.AppHostAssembly?.GetCustomAttributes<AssemblyMetadataAttribute>();
+        var objDir = GetMetadataValue(assemblyMetadata, "AppHostProjectBaseIntermediateOutputPath");
+
+        var fallbackDir = builder.Configuration[AspireStorePathKeyName];
+
+        var root = fallbackDir ?? objDir;
+
+        if (string.IsNullOrEmpty(root))
+        {
+            throw new InvalidOperationException($"Could not determine an appropriate location for storing user secrets. Set the {AspireStorePathKeyName} setting to a folder where the App Host content should be stored.");
+        }
+
+        var directoryPath = Path.Combine(root, ".aspire");
+
+        // The /obj directory doesn't need to be prefixed with the app host name.
+        if (root != objDir)
+        {
+            directoryPath = Path.Combine(directoryPath, GetAppHostSpecificPrefix(builder));
+        }
+
+        return new AspireStore(directoryPath);
+    }
+
+    /// <summary>
+    /// Gets the metadata value for the specified key from the assembly metadata.
+    /// </summary>
+    /// <param name="assemblyMetadata">The assembly metadata.</param>
+    /// <param name="key">The key to look for.</param>
+    /// <returns>The metadata value if found; otherwise, null.</returns>
+    private static string? GetMetadataValue(IEnumerable<AssemblyMetadataAttribute>? assemblyMetadata, string key) =>
+        assemblyMetadata?.FirstOrDefault(a => string.Equals(a.Key, key, StringComparison.OrdinalIgnoreCase))?.Value;
+
+    /// <summary>
+    /// Gets the application host specific prefix based on the builder's environment.
+    /// </summary>
+    /// <param name="builder">The <see cref="IDistributedApplicationBuilder"/>.</param>
+    /// <returns>The application host specific prefix.</returns>
+    private static string GetAppHostSpecificPrefix(IDistributedApplicationBuilder builder)
```

Comment:
I'm not sure we need this logic. I would be simplier if we just use the folder that is specified directly.

It is only used when the `Aspire:Store:Path` config is used. And in that case we can just tell people to use a different folder for each app, if they are having conflicts.